File: 2018/day6/solution1.py
from collections import Counter

def load_coordinates():
    coords = {}
    with open("input.txt") as fh:
        for i, line in enumerate(fh):
            line = line.rstrip('\n')
            x, y = line.split(', ')
            x, y = int(x), int(y)
            coords[i] = x, y
    return coords

def construct_grid(coords):
    grid = []
    max_x, max_y = 0, 0
    for x, y in coords.values():
        if x > max_x:
            max_x = x
        if y > max_y:
            max_y = y

    # We need to add 1 to the max x and y to account for the origin
    max_x += 1
    max_y += 1
    grid = [ [0] * max_x for i in range(max_y) ]

    return grid

def fill_grid(grid, coords):
    for row in range(len(grid)):
        for col in range(len(grid[0])):
            min_dist = len(grid) + len(grid[0])
            min_point = None
            mins = []
            for point in coords:
                dist = get_manhattan_dist((col, row), coords[point])
                if dist <= min_dist:
                    min_dist = dist
                    min_point = point
                    mins.append(min_dist)
            cmins = Counter(mins)
            if cmins[min_dist] > 1:
                grid[row][col] = -1
            else:
                grid[row][col] = min_point

def get_manhattan_dist(start, end):
    output = abs(start[0] - end[0]) + abs(start[1] - end[1])
    return output

def get_largest_area(grid):
    # Any point that is at the edge of the grid will be infinite so we create a
    # set of those point which we will ignore when getting the largest area
    infinite_set = set()
    for i in range(len(grid[0])):
        infinite_set.add(grid[0][i])
        infinite_set.add(grid[len(grid)-1][i])
    for j in range(len(grid)):
        infinite_set.add(grid[j][0])
        infinite_set.add(grid[j][len(grid[0])-1])

    count = Counter()
    for row in grid:
        count += Counter(row)

    largest_area = 0
    for point in count:
        if point != -1 and point not in infinite_set and count[point] > largest_area:
            largest_area = count[point]

    return largest_area


def main():
    coords = load_coordinates()
    grid = construct_grid(coords)
    fill_grid(grid, coords)
    largest_area = get_largest_area(grid)
    print("Size of the largest area: {}".format(largest_area))

if __name__ == '__main__':
    main()

File: 2018/day6/solution1.rs
use std::fs;
use std::io;
use std::io::BufRead;
use std::collections::HashMap;
use std::collections::HashSet;

#[derive(Debug)]
#[derive(Hash)]
#[derive(Copy, Clone)]
struct Point {
    x: i32,
    y: i32,
}

impl PartialEq for Point {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.y == other.y
    }
}

impl Eq for Point { }

fn construct_grid(coords: &Vec<Point>) -> Vec<Vec<Option<Point>>> {
    let mut grid: Vec<Vec<Option<Point>>> = Vec::new();
    let mut max_x = coords.iter().map(|p| p.x).max().unwrap();
    let mut max_y = coords.iter().map(|p| p.y).max().unwrap();

    // We need to add 1 to the max x and y to account for the origin
    max_x += 1;
    max_y += 1;
    for _ in 0..max_y {
        grid.push(vec![None;max_x as usize]);
    }

    grid
}

fn get_manhattan_dist(start: &Point, end: &Point) -> i32 {
    let output = i32::abs(start.x - end.x) + i32::abs(start.y - end.y);
    output
}

fn fill_grid(grid: &mut Vec<Vec<Option<Point>>>, coords: &Vec<Point>) {
    for row in 0..grid.len() {
        for col in 0..grid[0].len() {
            let cur_point = Point { x: col as i32, y: row as i32 };
            let mut min_dist: i32 = (grid.len() + grid[0].len()) as i32;
            let mut min_point = Point { x: 0, y: 0 };
            let mut mins = Vec::new();
            for p in coords.iter() {
                let dist = get_manhattan_dist(&cur_point, &p);
                if dist <= min_dist {
                    min_dist = dist;
                    min_point.x = p.x;
                    min_point.y = p.y;
                    mins.push(min_dist)
                }
            }
            let mut cmins = HashMap::new();
            for i in mins.iter() {
                let c = cmins.entry(i).or_insert(0);
                *c += 1;
            }
            if cmins[&min_dist] <= 1 {
                grid[row][col] = Some(min_point);
            }
        }
    }
}

fn load_coordinates() -> Vec<Point> {
    let mut output = Vec::new();

    let input = fs::File::open("input.txt")
        .expect("Something went wrong reading the file");
    let reader = io::BufReader::new(input);

    for line in reader.lines() {
        let line = line.unwrap().replace(" ", "");
        let coord_vec: Vec<&str> = line.split(',').collect();
        let x: i32 = coord_vec[0].parse().unwrap();
        let y: i32 = coord_vec[1].parse().unwrap();
        let p = Point { x: x, y: y };
        output.push(p);
    }
    output
}

fn get_largest_area(grid: Vec<Vec<Option<Point>>>) -> i32 {
    let mut infinite_set: HashSet<Point> = HashSet::new();
    for i in 0..grid[0].len() {
        if let Some(p) = &grid[0][i] {
            infinite_set.insert(*p);
        }
        if let Some(p) = &grid[grid.len()-1][i] {
            infinite_set.insert(*p);
        }
    }
    for j in 0..grid.len() {
        if let Some(p) = &grid[j][0] {
            infinite_set.insert(*p);
        }
        if let Some(p) = &grid[j][grid[0].len()-1] {
            infinite_set.insert(*p);
        }
    }

    let mut count = HashMap::new();
    for row in 0..grid.len() {
        for col in 0..grid[0].len() {
            if let Some(p) = grid[row][col] {
                if !infinite_set.contains(&p) {
                    let c = count.entry(p).or_insert(0);
                    *c += 1;
                }
            }
        }
    }

    *count.values().max().unwrap()
}

fn main() {
    let coords = load_coordinates();
    let mut grid = construct_grid(&coords);
    fill_grid(&mut grid, &coords);
    let largest_area = get_largest_area(grid);
    println!("Size of the largest area: {}", largest_area);
}

